pkg/controller/util/solr_util.go

```diff
@@ -234,6 +234,14 @@ func GenerateStatefulSet(solrCloud *solr.SolrCloud, ingressBaseDomain string, ho
 			VolumeClaimTemplates: pvcs,
 		},
 	}
+
+	if solrCloud.Spec.SolrImage.ImagePullSecret != "" {
+		log.Info("Adding imagePullSecret to statefulset", "ImagePullSecrets", solrCloud.Spec.SolrImage.ImagePullSecret)
+		stateful.Spec.Template.Spec.ImagePullSecrets = []corev1.LocalObjectReference{
```

Comment:
Thanks for pointing that out! Ok, let me checkout the CopyStatefulSetFields function and see if I can safely introduce this spec option there. 👍

Reply:
Ok, this spec comparison works as intended. Tested switching secret names and it updates statefulset and rolls out appropriately. The statefulset remains stable after update as well. Pods are **not** being recreated all of the time, so good to go there.

Reply:
Hmm, actually I see it being updated in the logs

{"level":"info","ts":1570727848.9402046,"logger":"controller","msg":"Adding imagePullSecret to statefulset","ImagePullSecrets":"example-secret-2"}

^ repeated a lot

But I don't see pods affected by this. Will look into this further. Not sure why the reflect detects a change every time when nothing as changed.

Reply:
I see now. It stops once the update completes though. It appears the statefulset has other values that differ while it is updating, perhaps it is the number of replicas fluctuates as the rollout proceeds. I can add some debugging logging to see what the difference is.

Once rolling update finishes the specs match up and reconcile process is happy.  The logs are quiet.

Reply:
With added logging, I can see that during a rolling update of statefulset the following reflect.DeeplyEqual checks are triggered for:

Spec.Template.Containers
Spec.Template.Spec.InitContainers
Spec.Template.Spec.Volumes

When I diff the output of both of these values they are the same. This behavior seems to have been present before this PR as a note. Just trying to dive into why https://github.com/bloomberg/solr-operator/blob/master/pkg/controller/util/solr_util.go#L281-L289 and https://github.com/bloomberg/solr-operator/blob/master/pkg/controller/util/solr_util.go#L296-L299 evaluate to true and require an updated only during  the statefulset rolling update process or when restarting the solr-operator pod.

If I rollback the changes made in this PR, I still see this behavior. I don't it will is causing any issues but might be related to the issues you ran into with Spec.Template.Spec.HostAliases 

I'll debug a bit further...

Reply:
Ill do some debugging with https://godoc.org/github.com/kylelemons/godebug/pretty#Compare

Reply:
Hmm, did some compares and there are some fields that appear to be changing in those portions of the StatefulSet. Must be some default values that get filled in during a recreation process that we aren't explicitly setting? Will share diff output here.

Reply:
The issue boils down to us generating a statefulset that is missing some fields that K8s fills in for us after submitting the statefulset. We then go into a reapply loop as we see the statefulset differ from what we expect. Ill include the fixes in this PR.

Reply:
InitContainer:
```
solr-operator-789748d947-kgn9z solr-operator -  TerminationMessagePath: "/dev/termination-log",
solr-operator-789748d947-kgn9z solr-operator -  TerminationMessagePolicy: "File",
solr-operator-789748d947-kgn9z solr-operator +  TerminationMessagePath: "",
solr-operator-789748d947-kgn9z solr-operator +  TerminationMessagePolicy: "",
```

Volumes:
```
solr-operator-c74bb688f-qc2v9 solr-operator       },
solr-operator-c74bb688f-qc2v9 solr-operator      ],
solr-operator-c74bb688f-qc2v9 solr-operator -    DefaultMode: 420,
solr-operator-c74bb688f-qc2v9 solr-operator +    DefaultMode: nil,
solr-operator-c74bb688f-qc2v9 solr-operator      Optional: nil,
solr-operator-c74bb688f-qc2v9 solr-operator     },
```

Container:
```
solr-operator-79c6c4cb54-74885 solr-operator -   TimeoutSeconds: 1,
solr-operator-79c6c4cb54-74885 solr-operator +   TimeoutSeconds: 0,
solr-operator-79c6c4cb54-74885 solr-operator     PeriodSeconds: 5,
solr-operator-79c6c4cb54-74885 solr-operator -   SuccessThreshold: 1,
solr-operator-79c6c4cb54-74885 solr-operator -   FailureThreshold: 3,
solr-operator-79c6c4cb54-74885 solr-operator +   SuccessThreshold: 0,
solr-operator-79c6c4cb54-74885 solr-operator +   FailureThreshold: 0,
solr-operator-79c6c4cb54-74885 solr-operator    },
solr-operator-79c6c4cb54-74885 solr-operator    Lifecycle: nil,
solr-operator-79c6c4cb54-74885 solr-operator -  TerminationMessagePath: "/dev/termination-log",
solr-operator-79c6c4cb54-74885 solr-operator -  TerminationMessagePolicy: "File",
solr-operator-79c6c4cb54-74885 solr-operator +  TerminationMessagePath: "",
solr-operator-79c6c4cb54-74885 solr-operator +  TerminationMessagePolicy: "",
...

solr-operator-79c6c4cb54-74885 solr-operator -   TimeoutSeconds: 1,
solr-operator-79c6c4cb54-74885 solr-operator +   TimeoutSeconds: 0,
solr-operator-79c6c4cb54-74885 solr-operator     PeriodSeconds: 10,
solr-operator-79c6c4cb54-74885 solr-operator -   SuccessThreshold: 1,
solr-operator-79c6c4cb54-74885 solr-operator -   FailureThreshold: 3,
solr-operator-79c6c4cb54-74885 solr-operator +   SuccessThreshold: 0,
solr-operator-79c6c4cb54-74885 solr-operator +   FailureThreshold: 0,
...

solr-operator-79c6c4cb54-74885 solr-operator      Name: "POD_HOSTNAME",
solr-operator-79c6c4cb54-74885 solr-operator      Value: "",
solr-operator-79c6c4cb54-74885 solr-operator      ValueFrom: {
solr-operator-79c6c4cb54-74885 solr-operator       FieldRef: {
solr-operator-79c6c4cb54-74885 solr-operator -      APIVersion: "v1",
solr-operator-79c6c4cb54-74885 solr-operator +      APIVersion: "",
...

solr-operator-79c6c4cb54-74885 solr-operator      Name: "solr-client",
solr-operator-79c6c4cb54-74885 solr-operator      HostPort: 0,
solr-operator-79c6c4cb54-74885 solr-operator      ContainerPort: 8983,
solr-operator-79c6c4cb54-74885 solr-operator -    Protocol: "TCP",
solr-operator-79c6c4cb54-74885 solr-operator +    Protocol: "",
```

adding fixes for these.

Reply:
Ok, commited fixes. Tested out locally and in development environment, statefulset creation and reconcile is stable :)

Reply:
Awesome debugging. Thanks so much for looking into this! It would be awesome if we could eventually get something more comprehensive than all of these if statements. But for now this should work.